code/extract_definition.py

```diff
@@ -0,0 +1,105 @@
+import re
+import os
+import html
+
+
+MULTIPLE_LINEBREAK_PATTERN = re.compile("\n+")
+FORMULA_PATTERN = re.compile(r'(Base|General) formula')
+DOLAR_PATTERN = re.compile(r'\$\$')
+ASTERIX_PATTERN = re.compile(r'\*\*')
+SENTENCE_WITH_PSET_PATTERN = re.compile(r'[^.?!]*\b(_Pset|_Qto)\b[^.?!]*[.?!]')
+HEADING_PATTERN = re.compile(r'\#+')
+KEYWORDS = ['NOTE', 'Note: ', 'DIAGRAM', 'CHANGE', 'IFC4', 'HISTORY',
+            'REFERENCE', 'EXAMPLE', 'DEPRECATION']
+
+
+def extract_definition(txt, return_short=True, return_marked=False, print_split=False):
+    """ Parse the original text and return only the semantic definition
+    or place a marker at its end. """
+
+    heading = txt.split('\n\n', 1)[0]+'\n\n'
+    txt = txt[len(heading):]
+    marker = '<!-- end of definition -->'
+
+    if not return_short and not return_marked:
+        return txt
+
+    try:
+        txt = html.unescape(txt)
+    except TypeError:
+        # Error when content is a link, for example:
+        # "https://github.com/buildingSMART/IFC4.3.x-development/edit/master/docs/schemas/core/
+        # IfcProductExtension/Types/IfcAlignmentTypeEnum.md#L0 has no content"
+        txt = ''
+    s1 = re.search(DOLAR_PATTERN, txt).start() if re.search(DOLAR_PATTERN, txt) else -1
+    s2 = re.search(FORMULA_PATTERN, txt).start() if re.search(FORMULA_PATTERN, txt) else -1
+    s3 = re.search(MULTIPLE_LINEBREAK_PATTERN, txt).start() if re.search(MULTIPLE_LINEBREAK_PATTERN,
+                                                                         txt) else -1
+    s4 = -1
+    for k in KEYWORDS:
+        x = txt.find(k)
+        if s4 == -1 and x != -1:
+            s4 = x
+        elif x < s4 and x != -1:
+            s4 = x
+    s5 = re.search(SENTENCE_WITH_PSET_PATTERN, txt).start() if re.search(SENTENCE_WITH_PSET_PATTERN,
+                                                                         txt) else -1
+    s6 = re.search(HEADING_PATTERN, txt).start() if re.search(HEADING_PATTERN, txt) else -1
+    i = min([x for x in [s1, s2, s3, s4, s5, s6, 1e5] if x >= 0])
+    if txt[i-1] in [":","-","–"]:
+        i = find_last_bullet_end_position(txt, i)
+
+    if txt[i - 1] == " ":
+        i = i - 1
+
+    # Execute
+    if i >= 0 and i != 1e5:
+        if print_split:
+            print('\n\n\033[92m'+heading + txt[:i] + '\033[91m' + txt[i:i+500]+'...\033[0m')
+        if return_short:
+            return txt[:i]
+        elif return_marked:
+            return heading + txt[:i] + marker + txt[i:]
+
+
+def find_last_bullet_end_position(text, i):
```

Comment:
Markdown is not something meant for parsing. What I would recommend instead: use markdown.markdown() to convert to HTML. Parse the HTML dom. And look for all list types: `ul` / `ol`. Your code does not consider numbered lists.

Reply:
That's a fair point. However, that is a single-use script that we are probably not going to use again (users will modify the markers manually from now on). I wondered if it should even be uploaded to the repo, but I decided we might need to use that split logic again.